### ISSUE TITLE
OpenCV DNN C++ throws exception with DeepLab MobileNet v2

### ISSUE THREAD

**Issue by baranorbert** [opencv/opencv#12715]
##### System information (version)
- OpenCV => OpenCV(4.0.0-pre) (latest version from GitHub repository)
- Operating System / Platform => Linux 64bit Ubuntu 16.04
- Compiler => gcc (5.4.0)
- Tensorflow => 1.10.1

##### Detailed description

I trained a Mobilnet v2 model with Tensorflow Deeplab. I used the default dataset (so I didn't replace the train images). The only thing what I modified is the "train_batch_size" parameter at training in the "local_test.sh" (original value: 4 => modified value: 1).

I tried to apply this model with the OpenCV DNN in C++, my test code was the following:
https://drive.google.com/file/d/1zM7T1FVNNxBPiM9QTH0pVdouAaF-ISFu/view?usp=sharing

I always got these errors:

- Error with the original or optimized frozen models:
OpenCV(4.0.0-pre) Error: The function/feature is not implemented (Unsupported squeeze configuration) in populateNet, file opencv/modules/dnn/src/tensorflow/tf_importer.cpp, line 1145
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.0.0-pre) opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1145: error: (-213:The function/feature is not implemented) Unsupported squeeze configuration in function 'populateNet'

- Error with the transformed frozen model:
OpenCV(4.0.0-pre) Error: Unspecified error (Tensor's data type is not supported) in getTensorContent, file /media/baran/B8703F33703EF828/baran/Devel/opencv/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp, line 773
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.0.0-pre) /media/baran/B8703F33703EF828/baran/Devel/opencv/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent
Aborted (core dumped)

##### Steps to reproduce

1. Training
Set "train_batch_size" to 1 in "tensorflow-models/research/deeplab/local_test.sh" and run this script.
Output: frozen_interference_graph.pb

2. Create proto file (frozen_interference_graph.pbtxt)
Apply exportpbtxt.py
https://drive.google.com/drive/folders/1dqhwObxvwuIWu1urwEZ0Qt8BjYx4THZr?usp=sharing

3. Optimize for inference
python tensorflow/tensorflow/python/tools/optimize_for_inference.py --input frozen_inference_graph.pb --output frozen_interference_graph_opt.pb --frozen_graph True --input_names ImageTensor --output_names SemanticPredictions
I always get this warning: "WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv/depthwise/BatchNorm/FusedBatchNorm'"

4. Create proto file (frozen_interference_graph_opt.pbtxt)
Apply exportpbtxt_opt.py
https://drive.google.com/drive/folders/1dqhwObxvwuIWu1urwEZ0Qt8BjYx4THZr?usp=sharing

5. Transform model
bazel run tensorflow/tools/graph_transforms:transform_graph -- --in_graph="frozen_inference_graph_opt.pb" --out_graph="frozen_inference_graph_transf.pb" --inputs="ImageTensor" --outputs="SemanticPredictions" --transforms="fold_constants sort_by_execution_order remove_nodes(op=Squeeze)"

6. Create proto file (frozen_interference_graph_transf.pbtxt)
Apply exportpbtxt_transf.py 
https://drive.google.com/drive/folders/1dqhwObxvwuIWu1urwEZ0Qt8BjYx4THZr?usp=sharing

7. Compile and execute C++ (Opencv dnn based) test app
Source code:
https://drive.google.com/file/d/1zM7T1FVNNxBPiM9QTH0pVdouAaF-ISFu/view?usp=sharing

Test image:
https://drive.google.com/file/d/1Wh3sK3CRLaHVTs0VrP0Wjc7v05x_FXrz/view?usp=sharing

**Comment by dkurt** [comment]
related: http://answers.opencv.org/question/199342/error-while-reading-deeplabv3-tensorflow-model-in-opencv/

**Comment by baranorbert** [comment]
Hi @dkurt, 

thanks the link, I have already saw this issue report. It covers only the first type of error: "Unsupported squeeze configuration in function 'populateNet'". But after I remove const and squeeze references with the transform script I got this error: "Tensor's data type is not supported in function 'getTensorContent'"

Do you need any further information or help to debug this issue?

Regards,
Norbert

**Comment by dkurt** [comment]
@baranorbert, try the following optimizations:
```python
import tensorflow as tf
from tensorflow.tools.graph_transforms import TransformGraph
from tensorflow.python.tools import optimize_for_inference_lib

graph = 'frozen_inference_graph.pb'
with tf.gfile.FastGFile(graph) as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    tf.summary.FileWriter('logs', graph_def)

    inp_node = 'MobilenetV2/MobilenetV2/input'
    out_node = 'logits/semantic/BiasAdd'
    graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, [inp_node], [out_node], tf.float32.as_datatype_enum)
    graph_def = TransformGraph(graph_def, [inp_node], [out_node], ["sort_by_execution_order"])
 
   with tf.gfile.FastGFile('frozen_inference_graph_opt.pb', 'wb') as f:
        f.write(graph_def.SerializeToString())
```

**Comment by baranorbert** [comment]
Hi @dkurt,

sorry for the late answer. I applied your script on the frozen_inference_graph.pb. I got some warning message, but the optimized Protobuf file was generated successfully and after that, I was able to read in and use it with OpenCV. Thank you very much your support! 

Warning messages:
WARNING:tensorflow:Passing a `GraphDef` to the SummaryWriter is deprecated. Pass a `Graph` object instead, such as `sess.graph`.
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_1/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_2/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_3/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_4/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_5/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_6/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_7/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_8/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_9/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_10/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_11/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_12/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_13/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_14/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_15/depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'MobilenetV2/expanded_conv_16/depthwise/BatchNorm/FusedBatchNorm'
2018-11-06 11:48:41.463416: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying sort_by_execution_order

**Comment by dkurt** [comment]
@baranorbert, That's good! The next step is to pass appropriate input to the network.

**Comment by baranorbert** [comment]
The script what @dkurt provided solved the problem so I can close this issue. Thanks a lot again!

**Comment by gwang92** [comment]
opencv version: 4.0.0
model :deeplab3+ mv2
Operating System : Linux 64bit Ubuntu 16.04

Detailed description
I tried to apply official model with the OpenCV DNN in C++,but I always got this error:
what(): OpenCV(4.0.0-rc) /home/crwon/opencv-4.0.0-rc/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent'

before this , I optimize for inference like you
According to dkurt commented,but still have this question:
what(): OpenCV(4.0.0-rc) /home/crwon/opencv-4.0.0-rc/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:773: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent'
